genomepy/caching.py

```diff
@@ -1,26 +1,87 @@
 import os
+import time
 from shutil import rmtree
+from functools import wraps
+from threading import Lock
 
 from appdirs import user_cache_dir
 from bucketcache import Bucket
-from joblib import Memory
+from joblib import Memory,dump,load,hash
+from diskcache import Cache
 
 from genomepy.__about__ import __version__
 
 genomepy_cache_dir = os.path.join(user_cache_dir("genomepy"), __version__)
 os.makedirs(genomepy_cache_dir, exist_ok=True)
-
 # Store the output of slow commands (marked with @cache and @goldfish_cache) for fast reuse.
 # Bucketcache creates a new pickle for each function + set of unique variables,
 # For class methods, use ignore=["self"] or use @staticmethod.
 cache = Bucket(genomepy_cache_dir, days=7)
 goldfish_cache = Bucket(genomepy_cache_dir, minutes=10)
 
 memory = Memory(genomepy_cache_dir, verbose=0)
-
+disk_cache = Cache(directory=genomepy_cache_dir)
 
 def clean():
     """Remove cached data on providers."""
     rmtree(genomepy_cache_dir, ignore_errors=True)
     os.makedirs(genomepy_cache_dir, exist_ok=True)
     print("All clean!")
+
+def hybcache(exp_short=3600, exp_long=3600, cache=disk_cache, lock=Lock()):
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            # Generate the cache key from the function's arguments.
+            print(args[0].__dict__)
+            key = [func.__name__] + list(args) + list(sorted(kwargs.items()))
+            #key = hash(key)
+            # Pickle object
+            ext = ".joblib"
+            pickle_obj = os.path.join(disk_cache.directory,f"{key}{ext}") #Check if value exists in the short term cache
+            value = cache.get(key)
+            #Check if the function call deos not exists in the memory cache
+            if value is None:
+                #Check if the value exists as pickle and if it is expired
+                with lock:
+                    if os.path.exists(pickle_obj):
+                        current_time = time.time()
+                        expiry_time = os.path.getctime(pickle_obj) + exp_long
+                        # Remove pickled object (long-term) if expired 
+                        if (current_time > expiry_time):
+                            print(f"Removing {key} from short-term memory since it is expired")
+                            os.remove(pickle_obj)
+                            # We need to re-compute and store the result in memory and as pickle
+                            print(f"re-computing {key} and caching value")
+                            value = func(*args, **kwargs)
+                            print(value)
+                            cache.close()
+                            with Cache(cache.directory) as reference:
```

Comment:
I _just_ googled this, so I might be wrong here. But cache comes from diskcache. This means you store the same file twice on disk. Just a FYI